### DISCUSSION TITLE
Impossible to install/compile Chia on FreeBSD

### DISCUSSION THREAD

**Discussion by Gherry777** [Chia-Network/chia-blockchain#9247] [General]
Hello,
today I tried to follow the instruction on how to install Chia on FreeBSD (I used a fresh install 11.4, even if it's outdated).

The instructions are not up to date and, in the end, even manually changing many variables it's impossible to install Chia.
The last error I got (after spending a couple of hours I gave up) is this one:

ERROR: Could not build wheels for chiapos, which is required to install pyproject.toml-based projects

Could anyone provide an updated guide (maybe for FreeBSD12) or could you just have an already compiled Chia if that is impossible (maybe a .pkg so it can be downloaded via the package manager).
Or please just remove support for FreeBSD and just tell people to use Ubuntu/Debian :(

Thank you in advance

**Comment by emlowe** [comment]
freebsd is mostly community-supported and the instructions on the wiki are also maintained by the community. Last we checked (which was in the last month) - they did work.

**Comment by mrrooster** [comment]
I have chia running in a jail on FreeBSD 13. 

Assuming a freshly created jail called 'chia' with internet access you first install the packages needed and create a chia user:

```bash
jexec chia pkg install bash git python screen bash gmake rust cmake gmp py38-sqlite3 py38-pybind11 pybind11
jexec chia bash -c '(echo chia::::::::: | adduser -f - -q -s bash -w no)'
```
Then enter the jail
```bash
jexec -U chia chia bash
```
Now cd to the home folder and start screen, I use a detached screen session to keep plotters running. You could omit this step, although you should probably still `cd` to ensure you're in the chia users folder:
```bash
cd 
screen
```

Next we build chia:
```bash
git clone https://github.com/Chia-Network/chia-blockchain.git -b latest
cd chia-blockchain
bash ./install.sh
```

This will fail, but has setup the python env we need, so we now install the missing clvm_rs:

```bash
. ./activate
cd ..
git clone https://github.com/Chia-Network/clvm_rs.git
cd clvm_rs
pip install maturin
maturin develop --release --cargo-extra-args="--features=openssl"
```
We can now run the build again:
```bash
cd ../chia-blockchain/
deactivate
bash ./install.sh
. ./activate
```
This should work second time round. Next `chia init`, etc. as usual. (NB: I don't run the gui).

One thing to note, if upgrading and you have a password protected keychain then edit `chia/util/keyring_wrapper.py` and change:
```
    elif platform == "linux":
```
to
```
    elif platform == "linux" or platform.startswith("freebsd"):
```
and put your keyring password in the string on the next line.

**Comment by cjedwards** [comment]
Confirmed that these instructions ever when followed to a T **_will not work_**
https://github.com/Chia-Network/chia-blockchain/wiki/FreeBSD-Install#clvm_rs
referencing python 3.7 which FreeBSD says NO, 
clvm does not install correctly. no cryptography directories are created to cp over etc.

I am trying mrrooster's suggestions above , I am running in the jails console so not using jexec,
So far I see that (py38-sqlite3 py38-pybind11 ) needs to be updated to 39
deactivate

I will post my results, hopefully positive, when done!

**Reply by cjedwards** [reply]
With only the small change noted above to  use 3.9 or 39 where required I was able to do a complete install with next to NO problems
It in fact did not even fail where suggested, 
I still did the manual install of clvm_rs

Thanks a million for the help.

Next up,, accessing 36 ZFS drives from jail!

**Comment by mrrooster** [comment]
I tried a fresh install recently and it's now pretty smooth on FreeBSD. If upgrading to 1.5.1 there's the possibility you could hit [this issue.](https://github.com/Chia-Network/chia-blockchain/issues/13282)

> Next up,, accessing 36 ZFS drives from jail!

I could never get the ZFS jail options to work the way I thought they should. (Not sure if it was my issue or they were genuinly broken).

My main chia has the ZFS storage mounted within it (by just changing the fs mount point.), however this is a home system and I do have user ID parity between my user and the chia user in the jail which saves a lot of permission messing about.

Your other option, and retrospectivly probably the one I would choose now is to use [nullfs](https://www.freebsd.org/cgi/man.cgi?query=nullfs&sektion=&n=1) to mount the ZFS fs into the jail.

You can do this with the `mount.fstab` jail option, for example, in your `jail.conf`:

```
myjail {
   ...
   mount.fstab=/etc/fstab.$name;
   ...
}
```
Create the mountpoint in your jail:
```sh
$ mkdir /usr/jail/myjail/mnt/chia_farm
```
Then create `/etc/fstab.myjail` and add something like:
```
/path/to/zfs   /usr/jail/myjail/mnt/chia_farm nullfs rw 0 0
```
Assuming your jail is rooted at `/usr/jail/myjail` you'll end up with `/path/to/zfs` mounted at `/mnt/chia_farm` in the jail.

**Reply by cjedwards** [reply]
I went right to 1.5.1 from scratch, so thankfully I did not encounter that problem.

"so far" using the UI to add Mount points to the Jailed chia instance seem to be working, and it allows me to correct some of the naming convention problems I created while figuring out how to track a bajillion single jbod drives.

It is labour intensive though, as in some cases I have to share twice from the same hdd, as I was learning how to use freebsd and truenas  from 0.  :) (moving 400+ TiB from NTFS → ZFS was a chore in itself)
